converters: use ftype in tests, add missing tests, lint tests

bernt-matthias opened the pull request:
While working on https://github.com/galaxyproject/galaxy/pull/12073 I realized that tests of many converters use the wrong attribute `format` instead of `ftype`.

`planemo l lib/galaxy/datatypes/converters/ --skip citations,stdio,help,tests,xml_order`  is now successful.

I would really like to add `planemo lint .. ` using `--skip` without `tests` to the `Converter tests`, but there are 39 converters without tests.

- [ ] interval_to_tabix
  - [ ] if gff is not included in the input types gff 2 bed converter is called first which leads to a diff
  - [ ] is it possible to incorporate vcf_bgzip_to_tabix in interval_to_tabix
  - [ ] gff_to_tabix is missing in datatypes.conf. should it be added?
- [ ] interval_to_bgzip: 
  - [ ] is it possible to include vcf to vcf_bgzip
- [ ] unify molecules converters
- [ ] unify converters using mdconvert

## How to test the changes?

- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bernt-matthias committed:
use ftype in converter tests

since format is wrong

nsoranzo committed:
Fix broken converter tests

bernt-matthias committed:
add tests to molecule converters

bernt-matthias committed:
add linting of converters and test results as artifact

- add skiplist to converter tests and lints

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

nsoranzo commented on .github/workflows/converter_tests.yaml:
```diff
@@ -1,34 +1,40 @@
+          name: Converter test results (${{ matrix.python-version }})
+          path: tool_test_output.html
+      - name: Lint converters
+        run: planemo lint 'galaxy root'/lib/galaxy/datatypes/converters/ --skip citations,stdio,help
```
I'd lint first and then test?

bernt-matthias committed:
add test for two more converters

- bz2_to_uncompressed
- cram_to_bam

bernt-matthias committed:
fix section order in gz_to_uncompressed

bernt-matthias committed:
cram_to_bam: add single quotes to tooldir

bernt-matthias committed:
add mock test for mothur datatype converter

bernt-matthias committed:
add test for picard interval list converter

nsoranzo commented on .github/workflows/converter_tests.yaml:
```diff
@@ -1,34 +1,45 @@
+      - name: Determine converters to check
+        run: |
+          ls 'galaxy root'/lib/galaxy/datatypes/converters/*xml | grep -v -f 'galaxy root'/lib/galaxy/datatypes/converters/.tt_skip | sed "s/^/'/; s/$/'/" > tool_list.txt
+          echo "skipping check for checking:"
```
```suggestion
          echo "Skipping checks for the following converters:"
```

bernt-matthias commented on .github/workflows/converter_tests.yaml:
```diff
@@ -1,34 +1,45 @@
+        run: |
+          ls 'galaxy root'/lib/galaxy/datatypes/converters/*xml | grep -v -f 'galaxy root'/lib/galaxy/datatypes/converters/.tt_skip | sed "s/^/'/; s/$/'/" > tool_list.txt
+          echo "skipping check for checking:"
+          ls 'galaxy root'/lib/galaxy/datatypes/converters/*xml | grep -f 'galaxy root'/lib/galaxy/datatypes/converters/.tt_skip 
```
```suggestion
          ls 'galaxy root'/lib/galaxy/datatypes/converters/*xml | grep -f 'galaxy root'/lib/galaxy/datatypes/converters/.tt_skip
          echo "Checking only the following converters:"
          cat tool_list.txt
```

bernt-matthias committed:
unify almost all converters to tabix

- all but vcf_bgzip

bernt-matthias committed:
converters: add test for neostorezip

bernt-matthias committed:
temporarily add test data

bernt-matthias committed:
mv test data

also to be removed before merging

nsoranzo commented on lib/galaxy/datatypes/converters/interval_to_bgzip_converter.xml:
```diff
@@ -12,15 +12,15 @@
     </command>
     <inputs>
-        <param format="interval" name="input1" type="data" label="Choose Interval file"/>
+        <param format="interval,encodepeak" name="input1" type="data" label="Choose Interval file"/>
```
This shouldn't be needed, since `ENCODEPeak` is a subclass of `Interval`.

bernt-matthias replied:
done

nsoranzo commented on test-data/encode.broad.peak:
```diff
@@ -0,0 +1,6 @@
+# TODO For the converter tests the header is commented fails otherwise
```
Should this and the following files go to https://github.com/galaxyproject/galaxy-test-data (when you're done)?

bernt-matthias replied:
Done. Lets see if this works

nsoranzo replied:
I don't see changes at https://github.com/galaxyproject/galaxy-test-data ?

nsoranzo commented on .github/workflows/converter_tests.yaml:
```diff
@@ -1,34 +1,65 @@
+        run: |
+          mapfile -t TOOL_ARRAY < tool_list.txt
+          for CONVERTER in "${TOOL_ARRAY[@]}"; do
+            planemo lint "$CONVERTER" --skip citations,stdio,help --report_level warn
```
```suggestion
            planemo lint --skip citations,stdio,help --report_level warn "$CONVERTER"
```

nsoranzo commented on .github/workflows/converter_tests.yaml:
```diff
@@ -1,34 +1,65 @@
+      - name: Lint converters
+        run: |
+          mapfile -t TOOL_ARRAY < tool_list.txt
+          for CONVERTER in "${TOOL_ARRAY[@]}"; do
```
I think it would be good to have an option in Planemo to take the list of tools to lint/test from a file (in addition to passing them directly as parameters on the command-line).
We would still need to merge reports when we run tests in parallel, but it would simplify the CI code nonetheless. Thoughts?

bernt-matthias commented on test-data/drugbank_drugs.smi:
```diff
@@ -0,0 +1 @@
+../lib/galaxy/datatypes/test/drugbank_drugs.smi
```
@nsoranzo are links a good idea here, or would you prefer to transfer them to https://github.com/galaxyproject/galaxy-test-data as well?

bernt-matthias replied:
Links seem to cause problems. I will finish this in a week.

nsoranzo replied:
Maybe move these files to test-data/ and symlink them back to `lib/galaxy/datatypes/test/` , that seems to be the current pattern.

bernt-matthias replied:
Done. Lets see if it works

bernt-matthias committed:
move test data

from lib/galaxy/datatypes/test to test-data

bernt-matthias committed:
unify to *_to_cml converters

nsoranzo commented on lib/galaxy/datatypes/converters/smi_to_cml_converter.xml:
```diff
@@ -1,6 +1,6 @@
 <tool id="CONVERTER_SMILES_to_cml" name="SMILES to CML" version="2.4.1">
```
Shouldn't this file be deleted now?

bernt-matthias replied:
Yes. Not sure yet about removing <parallelism>.. See my previous comment. Wanted to check if it works at all.

nsoranzo commented on lib/galaxy/datatypes/converters/sdf_to_cml_converter.xml:
```diff
@@ -1,6 +1,6 @@
 <tool id="CONVERTER_sdf_to_cml" name="SDF to CML" version="2.4.1">
```
Delete this file as well?

nsoranzo commented on lib/galaxy/datatypes/converters/inchi_to_cml_converter.xml:
```diff
@@ -1,6 +1,6 @@
 <tool id="CONVERTER_inchi_to_cml" name="InChI to CML" version="2.4.1">
```
Delete this file as well?

nsoranzo commented on lib/galaxy/datatypes/converters/mol_to_cml_converter.xml:
```diff
@@ -14,6 +14,19 @@
     <outputs>
```
Delete this file as well?

nsoranzo commented on lib/galaxy/datatypes/converters/mol2_to_cml_converter.xml:
```diff
@@ -1,6 +1,6 @@
 <tool id="CONVERTER_mol2_to_cml" name="MOL2 to CML" version="2.4.1">
```
Delete this file as well?

bernt-matthias committed:
move new test data to galaxy-test-data repo

bernt-matthias committed:
unify * to sdf converter

to replace the inchi, cml, mol2, smi to sdf converter

and to add the missing mol to sdf converter

bernt-matthias committed:
unify * to mol2 converters

bernt-matthias committed:
unify the * to inchi converters

bernt-matthias committed:
unify almost all to bgzip converters

- except vcf to vcf_bgzip

bernt-matthias committed:
test-data: move 2.cram and link to datasets test
